documentation/en/dev/architecture.md

```diff
@@ -0,0 +1,298 @@
+
+```sh
+lsof ~/.lotus/repo.lock
+# COMMAND   PID
+# lotus   52356
+```
+
+FIXME: Replace the `repo is already locked` error with the actual explanation so we don't need to also translate error to reality here as well: https://github.com/filecoin-project/lotus/issues/1829.
+
+The `node.Repo()` function (`node/builder.go`) contains most of the dependencies (specified as `Override()` calls) needed to properly set up the node's repo. We list the most salient ones here (defined in `node/modules/storage.go`).
+
+`LockedRepo()`: a common pattern in the DI model is to not only provide actual structures the node will incorporate in itself but list any type of dependencies in terms of "things that need to be done" before the node starts, this is the case of this function that doesn't create and initialize any new structure but rather registers a `OnStop` [hook](https://godoc.org/go.uber.org/fx/internal/lifecycle#Hook) that will close the locked repository associated to it:
+
+```Go
+type LockedRepo interface {
+	// Close closes repo and removes lock.
+	Close() error
+```
+
+`Datastore` and `ChainBlockstore`: all data related to the node state is saved in the repo's `Datastore`, an IPFS interface defined in `github.com/ipfs/go-datastore/datastore.go`. (See `(*fsLockedRepo).Datastore()` in `node/repo/fsrepo.go` for how Lotus creates it from a [Badger DB](https://github.com/dgraph-io/badger).) At the core every piece of data is a key-value pair in the `datastore` directory of the repo, but there are several abstractions we lay on top of it that appear through the code depending on *how* do we access it, but it is important to remember that the *where* is always the same.
+
+FIXME: Maybe mention the `Batching` interface as the developer will stumble upon it before reaching the `Datastore` one.
+
+The first abstraction is the `Blockstore` interface (`github.com/ipfs/go-ipfs-blockstore/blockstore.go`) which structures the key-value pair into the CID format for the key and the `Block` interface (`github.com/ipfs/go-block-format/blocks.go`) for the value, which is just a raw string of bytes addressed by its hash (which is included in the CID key/identifier). `ChainBlockstore` will create a `Blockstore` in the repo under the `/blocks` namespace (basically every key stored there will have that prefix so it does not collide with other stores that use the same underlying repo).
+
+FIXME: Link to IPFS documentation about DAG, CID, and related, especially we need a diagram that shows how do we wrap each datastore inside the next layer (datastore, batching, block store, gc, etc).
+
+Similarly, `modules.Datastore()` creates a `dtypes.MetadataDS` (alias for the basic `Datastore` interface) to store metadata under the `/metadata`. (FIXME: Explain *what* is metadata in contrast with the block store, namely we store the pointer to the heaviest chain, we might just link to that unwritten section here later.)
+
+FIXME: Explain the key store related calls.
```

Comment:
We can probably remove this, will describe the key store in a separate document explaining how to create a miner node.